data/ui_scripts/server_list/serverlist.lua

```diff
@@ -153,6 +336,192 @@ SystemLinkJoinMenu.AddServerButton = function(menu, controller, index)
+			side = "right",
+			clickable = false,
+			priority = -1000
+		}, function(f21_arg0, f21_arg1)
+			local type = Engine.GetDvarString("ui_server_type")
+			if type == "favourites" then
+				file = io.open("players2/favourites.txt", "w")
+				file:write("")
+				file:close()
+			elseif type == "blacklisted" then
+				file = io.open("players2/blacklisted.txt", "w")
+				file:write("")
+				file:close()
+			end
+			SystemLinkJoinMenu.BuildList(menu, Engine.GetFirstActiveController())
+		end)
+		
+		local servers = lines_as_hashmap("players2/favourites.txt")
+		for index = 1, #serversSorted, 1 do
+			local ip = Lobby.GetServerData(controller, serversSorted[index].index, 6)
+			if servers[ip] ~= nil then
+				SystemLinkJoinMenu.AddServerButton(menu, controller, serversSorted[index].index)
+			end 
+		end
+	elseif type == "blacklisted" then
+		menu.removefromblacklist = menu:AddHelp({
+			name = "add_button_helper_text",
+			button_ref = "button_right",
+			helper_text = Engine.Localize("@SERVERLIST_REMOVE_FROM_BLACKLIST"),
+			side = "right",
+			clickable = false,
+			priority = -1000
+		}, nil, nil, true)
+
+		menu.clearlist = menu:AddHelp({
+			name = "add_button_helper_text",
+			button_ref = "button_left_trigger",
+			helper_text = Engine.Localize("@SERVERLIST_CLEAR_LIST"),
+			side = "right",
+			clickable = false,
+			priority = -1000
+		}, function(f21_arg0, f21_arg1)
+			local type = Engine.GetDvarString("ui_server_type")
+			if type == "favourites" then
+				file = io.open("players2/favourites.txt", "w")
+				file:write("")
+				file:close()
+			elseif type == "blacklisted" then
+				file = io.open("players2/blacklisted.txt", "w")
+				file:write("")
+				file:close()
+			end
+			SystemLinkJoinMenu.BuildList(menu, Engine.GetFirstActiveController())
+		end)
+
+		local servers = lines_as_hashmap("players2/blacklisted.txt")
+		for index = 1, #serversSorted, 1 do
+			local ip = Lobby.GetServerData(controller,  serversSorted[index].index, 6)
+			if servers[ip] ~= nil then
+				SystemLinkJoinMenu.AddServerButton(menu, controller, serversSorted[index].index)
+			end 
+		end
+	else
+		menu.addtofavourites = menu:AddHelp({
+			name = "add_button_helper_text",
+			button_ref = "button_left",
+			helper_text = Engine.Localize("@SERVERLIST_ADD_TO_BLACKLIST"),
+			side = "right",
+			clickable = true,
+			priority = -1000
+		}, nil, nil, true)
+
+		menu.addtoblacklist = menu:AddHelp({
+			name = "add_button_helper_text",
+			button_ref = "button_right",
+			helper_text = Engine.Localize("@SERVERLIST_ADD_TO_FAVOURITES"),
+			side = "right",
+			clickable = true,
+			priority = -1000
+		}, nil, nil, true)
+	
+		local servers = lines_as_hashmap("players2/blacklisted.txt")
+		for index = 1, #serversSorted, 1 do
+			local ip = Lobby.GetServerData(controller,  serversSorted[index].index, 6)
+			if servers[ip] == nil then
+				SystemLinkJoinMenu.AddServerButton(menu, controller, serversSorted[index].index)
+			end
+		end
```

Comment:
This should also be done in C++ because it's still pinging all the servers even if they are just going to be filtered out which is wasteful, not to mention that if a favorite server isn't on the master server list it won't appear

Reply:
You should make it so the servers are fetched from the favorites list instead of the master server so it only pings the servers it needs, then you should add a builtin lua function to switch between master/favorites/blacklisted, same thing goes for the sorting

Reply:
I have no idea on how to make this kind of things.